klippy/extras/display/menu.py

```diff
@@ -0,0 +1,1137 @@
+
+        self.running = False
+
+    def update_parameters(self, eventtime):
+        self.parameters = {
+            'screen': {
+                'eventtime': eventtime,
+                'is2004': (self.rows == 4 and self.cols == 20),
+                'is2002': (self.rows == 2 and self.cols == 20),
+                'is1604': (self.rows == 4 and self.cols == 16),
+                'is1602': (self.rows == 2 and self.cols == 16)
+            }
+        }
+        for name in self.objs.keys():
+            if self.objs[name] is not None and type(self.objs[name]) != dict:
+                try:
+                    self.parameters[name] = self.objs[name].get_status(eventtime)
+                except:
+                    self.parameters[name] = {}
+                self.parameters[name].update({'is_enabled': True})
+                # get additional info
+                if name == 'toolhead':
+                    pos = self.objs[name].get_position()
+                    self.parameters[name].update({'xpos':pos[0], 'ypos':pos[1], 'zpos':pos[2], 'epos': pos[3]})
+                    self.parameters[name].update({
+                        'is_printing': (self.parameters[name]['status'] == "Printing"),
+                        'is_ready': (self.parameters[name]['status'] == "Ready"),
+                        'is_idle': (self.parameters[name]['status'] == "Idle")
+                    })
+                elif name == 'extruder0':
+                    info = self.objs[name].get_heater().get_status(eventtime)
+                    self.parameters[name].update(info)
+                elif name == 'extruder1':
+                    info = self.objs[name].get_heater().get_status(eventtime)
+                    self.parameters[name].update(info)
+                elif name == 'display':
+                    self.parameters[name].update({
+                        'progress': 0 if self.objs[name].progress is None else self.objs[name].progress,
+                        'progress.visible': not not self.objs[name].progress,
+                        'message': '' if self.objs[name].message is None else str(self.objs[name].message),
+                        'message.visible': not not self.objs[name].message,
+                        'is_enabled': True
+                    })
+            elif type(self.objs[name]) == dict:
+                self.parameters[name] = {}
+                if name == 'output_pin' or name == 'servo':
+                    for key, obj in self.objs[name].items():
+                        try:
+                            self.parameters[name].update({'%s.value' % str(key): obj.last_value, '%s.is_enabled' % str(key): True})
+                        except Exception as e:
+                            logging.error('Parameter update error: '+ str(e))
+                else:
+                    self.parameters[name].update({'is_enabled': False})
+            else:
+                self.parameters[name] = {'is_enabled': False}
```

Comment:
I think update_parameters() could become cumbersome to maintain.  Could we instead enhance `get_status()` in the individual modules to return the desired values?

To be clear, lets not do that as part of this commit, as the change is already large.  At a high-level though, I think we'll need to figure out a way to have the other printer modules list what display capabilities they support and have the menu code automatically find all modules with that display support.  I fear the other method (menu knowing about all the other modules) will result in lots of code churn in the menu.